src/core/net/ip6.cpp

```diff
@@ -1201,56 +1201,38 @@ Error Ip6::HandleDatagram(Message &aMessage, MessageOrigin aOrigin, const void *
         }
     }
 
-    // Never forward reassembled frames as they were already delivered
-    // as fragments.
-    if (aIsReassembled)
-    {
-        forwardHost = false;
-    }
-
     aMessage.SetOffset(sizeof(header));
 
     // Process IPv6 Extension Headers
     nextHeader = static_cast<uint8_t>(header.GetNextHeader());
     SuccessOrExit(error = HandleExtensionHeaders(aMessage, aOrigin, messageInfo, header, nextHeader, receive));
 
-    // Process IPv6 Payload
-    if (receive)
+    if (receive && (nextHeader == kProtoIp6))
     {
-        if (nextHeader == kProtoIp6)
-        {
-            // Remove encapsulating header and start over.
-            aMessage.RemoveHeader(aMessage.GetOffset());
-            Get<MeshForwarder>().LogMessage(MeshForwarder::kMessageReceive, aMessage);
-            goto start;
-        }
-
-        if (!aIsReassembled)
-        {
-            error = PassToHost(aMessage, aOrigin, messageInfo, nextHeader,
-                               /* aApplyFilter */ !forwardHost, Message::kCopyToUse);
-
-            if ((error == kErrorNone || error == kErrorNoRoute) && forwardHost)
-            {
-                forwardHost = false;
-            }
-        }
+        // Remove encapsulating header and start over.
+        aMessage.RemoveHeader(aMessage.GetOffset());
+        Get<MeshForwarder>().LogMessage(MeshForwarder::kMessageReceive, aMessage);
+        goto start;
+    }
 
-        error = HandlePayload(header, aMessage, messageInfo, nextHeader,
-                              (forwardThread || forwardHost ? Message::kCopyToUse : Message::kTakeCustody));
+    if ((forwardHost || receive) && !aIsReassembled)
+    {
+        error = PassToHost(aMessage, aOrigin, messageInfo, nextHeader,
+                           /* aApplyFilter */ !forwardHost,
+                           (receive || forwardThread) ? Message::kCopyToUse : Message::kTakeCustody);
 
         // Need to free the message if we did not pass its
-        // ownership in the call to `HandlePayload()`
-        shouldFreeMessage = forwardThread || forwardHost;
+        // ownership in the call to `PassToHost()`
+        shouldFreeMessage = (receive || forwardThread);
```

Comment:
You suggest we add a `shouldFreeMessage = forwardThread` after the `if (receive)` block.
I am not sure if we can combine them. 

I think it won't work when we have all three booleans as `false` (i.e., `forwardHost == false`, `receive == false`, and `forwardThread == false`). In this case, `shouldFreeMessage` will be set to `false` but we should still free the message.

---

The way we use and set `shouldFreeMessage` is kinda of tricky. 
- We set `shouldFreeMessage = true` at the top (basically we always want to free the message).
- Only when we call `PassToHost()` and `HandlePayload` and use `kTakeCustody` we pass ownership of `Message` (and we expect the method we called to free it), we need to set the `shouldFreeMessage` to `false` to avoid double free.

I plan/hope to later change and simplify this. The idea is to use `Message *` and when we pass ownership we set the pointer to `nullptr` - at the end we free if it is not `nullptr`.

Reply:
That makes sense. I missed that `shouldFreeMessage` is initialized to `true`.